### ISSUE TITLE
Modify cosine angle functional form

### ISSUE THREAD

**Issue by proteneer** [proteneer/timemachine#743]
Our current harmonic angle oscillator has an option for `cos_angles` that we default to `True`. The original motivation for this was to avoid the numerical singularity that arises at 0 or 180 degrees. Most MD packages patch this behavior in odd ways, but it's really undesirable. So when I initially wrote the angle term, I adopted the gromos functional form of (cos(a)-cos(t))^2, which while it solves the issue at 0 and pi, it fails spectacularly for other values. After some debugging, (cos(a-t)-1)^2 is a much better approximation:

(red: original, green: gromos, current timemachine, purple: proposed)
![Screen Shot 2022-05-17 at 6 07 41 PM](https://user-images.githubusercontent.com/2280724/168918744-590724d7-6a4d-4041-867b-0ecdb55c8b7d.png)
![Screen Shot 2022-05-17 at 6 07 49 PM](https://user-images.githubusercontent.com/2280724/168918747-e3b64ea4-97f9-4a3d-ae30-e63c7ecfae0f.png)
![Screen Shot 2022-05-17 at 6 07 56 PM](https://user-images.githubusercontent.com/2280724/168918749-e57793e4-87e9-4588-b061-3f7bd6412266.png)
![Screen Shot 2022-05-17 at 6 08 08 PM](https://user-images.githubusercontent.com/2280724/168918751-05fb6b56-03e3-4bd3-90c8-34d8f56571de.png)

**Comment by proteneer** [comment]
In
<img width="1581" alt="Screen Shot 2022-05-17 at 9 13 42 PM" src="https://user-images.githubusercontent.com/2280724/168937154-71a3e40b-4476-4002-a91f-c8046a6b4908.png">
terestingly enough, expressed as cartesian coordinates, the gromos form actually fits quite well, but the barriers may need to be rescaled.

**Comment by proteneer** [comment]
I think the gromos forcefield is actually fine, since the use of `y=arccos(x)`, where `-1<=x<=1` restricts the output to `0<=y<=pi`:
<img width="1523" alt="Screen Shot 2022-05-17 at 9 23 32 PM" src="https://user-images.githubusercontent.com/2280724/168938379-ed4690c6-4e1d-4b83-bca0-89ff8edb54a4.png">

**Comment by proteneer** [comment]
TL;DR: we should refit the force constants here with wrap-around considerations.

**Comment by proteneer** [comment]
Note: a good zero-th order approximation is probably to just increase k by a factor of 2.

**Comment by maxentile** [comment]
Clarification: is the main concern:
1. force singularities that occur at a single point (dU/dx = nans when theta = theta_0 = 0 for example),
2. energy discontinuities due to non-periodic angular distance definition (theta - theta_0)^2

?

Possible workaround for 2: replace direct `delta_theta = theta - theta_0` with periodic version:

<img width="636" alt="image" src="https://user-images.githubusercontent.com/5759036/169060015-5ac9747a-a53d-45bd-ad50-9d579e158211.png">


```python

import numpy as np
from jax import numpy as jnp
import matplotlib.pyplot as plt

theta_grid = np.linspace(-2 * np.pi, 2 * np.pi, 10000)

def u_direct(theta, theta_0=0, k=1.0):
    return 0.5 * k * (theta - theta_0)**2

def direct_delta_theta(theta, theta_0):
    """does not satisfy d(theta, theta_0) = d(theta + 2pi, theta_0)"""
    return theta - theta_0

def periodic_delta_theta(theta, theta_0):
    """by analogy to PBC-compatible definition of delta_r"""
    # https://github.com/proteneer/timemachine/blob/451803e01afe6231147a0e6a3ca019d4aa5069d8/timemachine/potentials/jax_utils.py#L70-L77
    diff = theta - theta_0
    width = 2 * np.pi
    return diff - (width * jnp.floor(diff / width + 0.5))

def u_direct(theta, theta_0=0, k=1.0):
    return 0.5 * k * direct_delta_theta(theta, theta_0)**2

def u_periodic(theta, theta_0=0, k=1.0):
    return 0.5 * k * pbc_delta_theta(theta, theta_0)**2

plt.figure(figsize=(8,3))
ax = None
for i, theta_0 in enumerate([0, np.pi, 2*np.pi]):
    ax = plt.subplot(1, 3, i+1, sharey=ax)
    plt.title(rf'$\theta_0$ = {theta_0/np.pi} $\pi$')
    #theta_grid = np.linspace(theta_0 - 2 * np.pi, theta_0 + 2* np.pi, 10000)
    plt.plot(theta_grid, u_direct(theta_grid, theta_0=theta_0), label='direct')
    plt.plot(theta_grid, u_periodic(theta_grid, theta_0=theta_0), label='periodic')
    plt.xlabel(r'$\theta$'); plt.ylabel(r'$U(\theta)$')
    if i == 0: plt.legend()
plt.tight_layout()
```

Doesn't solve the problem of the forces being discontinuous at `theta_0 +/- pi`, but this might be of limited practical concern when `k` is large.

**Comment by maxentile** [comment]
(Tangent: May be worth looking at numerical stability considerations for other angular distributions such as the Von Mises distribution, e.g. in Stan https://mc-stan.org/docs/2_22/functions-reference/von-mises-distribution.html or Mitsuba https://www.mitsuba-renderer.org/devblog/2012/07/numerically-stable-sampling-of-the-von-mises-fisher-distribution-on-s2-and-other-tricks/ )

**Comment by proteneer** [comment]
The main concern is:

1) force singularities that occur at a single point (dU/dx = nans when theta = theta_0 = 0 for example),

There are no discontinuities in the energy in any of the above functional forms.

**Comment by maxentile** [comment]
Closed by https://github.com/proteneer/timemachine/pull/1271
